[Ukrainian] Translate `/advanced/deployment.md` (#698)
grammyjs/website: site/docs/uk/advanced/deployment.md

---
prev: ./proxy.md
---

# Контрольний список розгортання

Ось список речей, які слід памʼятати при розміщенні на хостингу великого бота.

> Вас також можуть зацікавити наші інструкції по розміщенню ботів.
> Перегляньте **Хостинг/Посібники** у верхній частині сторінки, щоб побачити деякі платформи, які вже мають спеціальні посібники.

## Помилки

1. [Встановіть обробник помилок за допомогою `bot.catch` для тривалого опитування або у вашому серверному фреймворку для вебхуків.](../guide/errors.md)

> 2. Використовуйте `await` для всіх `Promise` і встановіть **лінтинг** з правилами, які забезпечують це, щоб ви ніколи не забували.

Reviewer: ```suggestion
2. Використовуйте `await` для всіх `Promise` і встановіть **лінтинг** з правилами, які забезпечують це, щоб ви ніколи про це не забували.
```

## Надсилання повідомлень

> 1. Надсилайте файли, передаючи шлях або `Buffer` замість потоку, або принаймні переконайтеся, що ви [знаєте підводні камені](./transformers.md#випадки-використання-перетворювачів).

Reviewer: ```suggestion
1. Надсилайте файли, передаючи шлях або `Buffer` замість потоку, або принаймні переконайтеся, що ви [знаєте про можливі підводні камені](./transformers.md#випадки-використання-перетворювачів).
```

2. Використовуйте `bot.on("callback_query:data")` як резервний обробник для [реагування на всі запити зворотного виклику](../plugins/keyboard.md#відповідь-на-натискання).

> 3. Використовуйте [плагін `transformer-throttler`](../plugins/transformer-throttler.md), щоб уникнути перевищення ліміту швидкості.

Reviewer: ```suggestion
3. Використовуйте [плагін `transformer-throttler`](../plugins/transformer-throttler.md), щоб уникнути перевищення ліміту кількості запитів.
```

> 4. **Необовʼязково**. Розгляньте можливість використання [плагіна `auto-retry`](../plugins/auto-retry.md) для автоматичного оброблення помилок, викликаних невдалими запитами, та повторного надсилання запиту.

Reviewer: ```suggestion
4. **Необовʼязково**. Розгляньте можливість використання [плагіна `auto-retry`](../plugins/auto-retry.md) для автоматичного оброблення помилок, викликаних перевищенням ліміту кількості запитів, та повторного надсилання таких запитів.
```

## Масштабування

Це залежить від типу розгортання.

> ### Тривале опитування

Reviewer: ```suggestion
### Тривале опитування (long polling)
```

Author: Нащо тут це уточнення? Думаю, читач спочатку прочитає про ці види роботи бота в посібнику, тому вже розрізнятиме їх.

Reviewer 2: I'm agree with @deptyped.

> 1. [Використовуєте плагін для конкурентності (runner).](../plugins/runner.md)

Reviewer 2: ```suggestion
1. [Використовуйте плагін для конкурентності (runner).](../plugins/runner.md)
```

> 2. [Використовуйте `sequentialize` з тією самою функцією вирішувача ключів сеансів, що і middleware сеансу.](./scaling.md#паралелізм-важкий)

Reviewer: ```suggestion
2. [Використовуйте `sequentialize` з тією самою функцією для отримання ключу сесії, що і в плагіні сесій.](./scaling.md#паралелізм-важкий)
```

> 3. Перегляньте параметри конфігурації `run` ([довідка API](https://deno.land/x/grammy_runner/mod.ts?s=run)) і переконайтеся, що вони відповідають вашим потребам, або навіть розгляньте можливість створення власного бігуна з [джерелами](https://deno.land/x/grammy_runner/mod.ts?s=UpdateSource) та [поглиначами](https://deno.land/x/grammy_runner/mod.ts?s=UpdateSink).

Author: Залишу це тут до тих пір, доки не прийде час review
```suggestion
3. Перегляньте параметри конфігурації `run` ([довідка API](https://deno.land/x/grammy_runner/mod.ts?s=run)) і переконайтеся, що вони відповідають вашим потребам, або навіть розгляньте можливість створення власного runner'а з [джерелами (source)](https://deno.land/x/grammy_runner/mod.ts?s=UpdateSource) та [поглиначами (sink)](https://deno.land/x/grammy_runner/mod.ts?s=UpdateSink).
```

Reviewer: ```suggestion
3. Перегляньте параметри конфігурації `run` ([довідка API](https://deno.land/x/grammy_runner/mod.ts?s=run)) і переконайтеся, що вони відповідають вашим потребам, або навіть розгляньте можливість зібрати власний runner зі своїм [джерелом оновлень](https://deno.land/x/grammy_runner/mod.ts?s=UpdateSource) та їх [обробником](https://deno.land/x/grammy_runner/mod.ts?s=UpdateSink).
```

Author: Точно не треба залишити тут оригінальний термін? Ми ж наче обговорювали, що при першій згадці на сторінці залишимо оригінальний термін поруч з перекладом

   Головне, що потрібно враховувати - це максимальне навантаження, яке ви хочете застосувати до вашого сервера, тобто скільки оновлень може бути оброблено одночасно.

> 4. Подумайте про реалізацію [graceful shutdown](./reliability.md#graceful-shutdown), щоб зупинити бота, коли ви захочете його вимкнути, тобто перейти на нову версію.

Reviewer: ```suggestion
4. Подумайте про реалізацію [graceful shutdown](./reliability.md#graceful-shutdown), щоб завершити обробку поточних оновлень перш ніж ви зупините бота, наприклад, для його оновлення.
```

### Вебхуки

1. Переконайтеся, що ви не виконуєте довготривалі операції у middleware: наприклад, передача великих файлів.
   [Це призводить до помилок тайм-ауту](../guide/deployment-types.md#своєчасне-завершення-запитів-вебхуків) для вебхуків та дублювання обробки оновлень, оскільки Telegram буде повторно надсилати непідтверджені оновлення.

>    Замість цього розгляньте можливість використання системи черг завдань.

Reviewer: ```suggestion
   Замість цього розгляньте можливість використання черги завдань.
```

2. Ознайомтеся з конфігурацією `webhookCallback` ([довідка API](https://deno.land/x/grammy/mod.ts?s=webhookCallback)).

> 3. Якщо ви налаштували параметр `getSessionKey` для вашого сеансу, [використовуйте `sequentialize` з тією ж функцією вирішувача ключів сеансів, що і middleware сеансу](./scaling.md#паралелізм-важкий).

Reviewer: ```suggestion
3. Якщо ви змінювали параметр `getSessionKey` для плагіну сесій, [використовуйте `sequentialize` з тією ж функцією для отримання ключу сесії](./scaling.md#паралелізм-важкий).
```

4. Якщо ви працюєте на безсерверній платформі або платформі з автоматичним масштабуванням, [встановіть інформацію про бота](https://deno.land/x/grammy/mod.ts?s=BotConfig), щоб запобігти надмірній кількості викликів `getMe`.
5. Подумайте про використання [відповідей вебхуку](../guide/deployment-types.md#відповідь-вебхуку).

## Сесії

1. Подумайте про використання `lazySessions`, як описано [тут](../plugins/session.md#ліниві-сесії).
2. Використовуйте опцію `storage` для налаштування вашого адаптера зберігання даних, інакше всі дані будуть втрачені, коли процес бота зупиниться.

## Тестування

Пишіть тести для свого бота.
Це можна зробити за допомогою grammY наступним чином:

1. Імітуйте вихідні запити API за допомогою [перетворювачів](./transformers.md).
2. Визначте та надішліть зразки обʼєктів оновлення вашому боту за допомогою `bot.handleUpdate` ([довідка API](https://deno.land/x/grammy/mod.ts?s=Bot#method_handleUpdate_0)).

>    Подумайте про те, щоб надихнутися цими [обʼєктами оновлення](https://core.telegram.org/bots/webhooks#testing-your-bot-with-updates), наданими командою Telegram.

Reviewer: ```suggestion
   Скористайтеся цими [обʼєктами оновлення](https://core.telegram.org/bots/webhooks#testing-your-bot-with-updates), наданими командою Telegram, для натхнення.
```
